### ISSUE TITLE
Suggestion for 0. Entries and 1. Hashes: Clarify entry hash location and utility

### ISSUE THREAD

**Issue by pyramid-songs** [holochain-gym/holochain-gym.github.io#29]
Re-reading through 0. and 1.

I had a bit of an 'aha' moment looking through the contents of `Element` (inspired by your [discussion here](https://github.com/holochain-gym/holochain-gym.github.io/issues/24#issuecomment-817312497)). Entries themselves do not contain the hash of the entry; that is stored as part of SignedHeaderHashed, and it's why we need to perform the hash_entry(struct) in the first place so we don't have to trek through the following piece of code just to get the entry hash:

`element.into_inner().0.header().entry_hash().unwrap().clone()`

This is obvious to the three of us now, but imagine a new reader seeing the following in the first exercise:

```
In the Entry Contents you see three things:

    `Entry Hash`...
    `entry_type`...
    `content`...
```

They would most likely think that the entry has the hash in it, even though looking into the future exercises it becomes obvious that the entry hash has to be found externally from the entry, unless you hash it yourself.

Addressing this early on in the learning process would be good, ideally in chapter 1.Hashes, the Getting Ready section nearby this: 

> In one of the next exercise we will see that there are other ways to retrieve your data: headers and elements. But it is good to know you can always find your entry again based on its entry hash.

Maybe we should add something to highlight what I mentioned above: that `hash_entry()` is the preferred way to find entry hashes. This kind of begs the question of why doesn't `create_entry()` return a tuple of HeaderHash and EntryHash, which I supposed can be answered: the entry hash is already in the header portion of the element, and we should use a map that points to the treasure rather than try to find the treasure using just a picture of the treasure chest itself. This is hinted at in the quoted section.

TL;DR - I think we give a conflicting impression of entry hashes in chapters 0 and 1 that needs to be tightened a little.

**Comment by guillemcordoba** [comment]
Hi @pyramid-songs ! I see what you are getting at. The `Entry Contents` UI block puts the hash first because it is the ID for the entry/header, and then you have the json blob of data which are their contents. It has always felt weird to me to have it there, maybe we add it in the title? As in `Entry Contents, for entry ufsdlkfsdjfl`, although that may be too long...

I would hope that the hashes exercise gives a bit of practice for 2 things: 1) when hashing something, the hash can never be part of the content we are hashing (because then that would change the hash, and so on in an infinite loop), and 2) if you have the contents, you can derive its hash always, even when that entry doesn't even exist - it is just a mathematical function after all. Maybe in the section for anchors we can exercise this more. Maybe we need to state this explicitly?

`create_entry` returned an `EntryHash` in the past version of holochain (redux), but now in rsm it returns the header hash because elements have become much more important. As a practical reason, it's easy to get the hash of the entry from its contents, but to get the latest committed header hash one has to do a query pass for the source chain of the agent, which wouldn't be that performant.

Anyway, I think your aha moment was great, and those happen at different times for different people. Do you wanna propose concrete changes to those exercises?

**Comment by pyramid-songs** [comment]
That all makes sense to me! Looking back on it, I think my initial confusion came from assumptions I made about the structure of the entry due to the UI block and the source chain graphs. 

Happy to make those changes. I'll have a think about what specifically needs to change and how to phrase it.

**Comment by guillemcordoba** [comment]
I'll play around with the `Entry Contents` block and see how we can place the hash better.
